planner/core/hints.go

```diff
@@ -22,7 +22,44 @@ import (
 	utilhint "github.com/pingcap/tidb/util/hint"
 )
 
+// GenHintsFromFlatPlan generates hints from a FlatPhysicalPlan.
+func GenHintsFromFlatPlan(flat *FlatPhysicalPlan) []*ast.TableOptimizerHint {
+	if len(flat.Main) == 0 {
+		return nil
+	}
+	nodeTp := utilhint.TypeSelect
+	switch flat.Main[0].Origin.(type) {
+	case *Update:
+		nodeTp = utilhint.TypeUpdate
+	case *Delete:
+		nodeTp = utilhint.TypeDelete
+	}
+	var hints []*ast.TableOptimizerHint
+	selectPlan := flat.Main.GetSelectPlan()
+	if len(selectPlan) == 0 || !selectPlan[0].IsPhysicalPlan {
+		return nil
+	}
+	for _, op := range selectPlan {
+		if !op.IsRoot {
+			continue
+		}
+		p := op.Origin.(PhysicalPlan)
+		hints = genHintsFromSingle(p, nodeTp, hints)
+	}
+	for _, cte := range flat.CTEs {
+		for i, op := range cte {
+			if i == 0 || !op.IsRoot {
+				continue
+			}
+			p := op.Origin.(PhysicalPlan)
+			hints = genHintsFromSingle(p, nodeTp, hints)
+		}
+	}
+	return hints
+}
+
 // GenHintsFromPhysicalPlan generates hints from physical plan.
+// Deprecated: FlattenPhysicalPlan() + GenHintsFromFlatPlan() is preferred.
 func GenHintsFromPhysicalPlan(p Plan) []*ast.TableOptimizerHint {
```

Comment:
The only place it is called in prod env is in the `Optimize`'s SPM evolve phase. In that place, we get the hints from the plan returned by `optmize()`. Would we also flatten the plan there?